postgre-client/src/tasks.rs

```diff
@@ -93,51 +169,106 @@ impl PgClient {
         });
 
-        query_builder.push(") as tmp (tsk_metadata_url, tsk_status, tsk_error) WHERE tasks.tsk_metadata_url = tmp.tsk_metadata_url;");
+        query_builder.push(") as tmp (tasks_metadata_hash, tasks_metadata_url, tasks_task_status, tasks_etag, tasks_last_modified_at, tasks_mutability, tasks_next_try_at, tasks_error) WHERE tasks.tasks_metadata_hash = tmp.tasks_metadata_hash;");
 
         let query = query_builder.build();
         query.execute(&self.pool).await?;
 
         Ok(())
     }
 
-    pub async fn get_pending_tasks(
+    pub async fn update_tasks_attempt_time(&self, data: Vec<String>) -> Result<(), IndexDbError> {
+        let mut query_builder: QueryBuilder<'_, Postgres> = QueryBuilder::new(
+            "
+            UPDATE tasks
+            SET
+                tasks_next_try_at = NOW() + INTERVAL '1 day'
+                WHERE tasks.tasks_metadata_url IN (
+        ",
+        );
+
+        query_builder.push_values(data, |mut b, task| {
+            let url = UrlWithStatus::new(task.as_str(), false); // status is ignored here
+            b.push_bind(url.metadata_url);
+        });
+
+        query_builder.push(");");
+
+        let query = query_builder.build();
+        query.execute(&self.pool).await?;
+
+        Ok(())
+    }
+
+    pub async fn get_pending_metadata_tasks(
         &self,
         tasks_count: i32,
-    ) -> Result<Vec<JsonDownloadTask>, IndexDbError> {
-        let mut query_builder: QueryBuilder<'_, Postgres> =
-            QueryBuilder::new("WITH cte AS (
-                                        SELECT tsk_id
-                                        FROM tasks
-                                        WHERE (tsk_status = 'running' OR tsk_status = 'pending') AND tsk_locked_until < NOW() AND tsk_attempts < tsk_max_attempts
-                                        LIMIT ");
-
+    ) -> Result<Vec<MetadataDownloadTask>, IndexDbError> {
+        // skip locked not to intersect with synchronizer work
+        let mut query_builder: QueryBuilder<'_, Postgres> = QueryBuilder::new(
+            "WITH selected_tasks AS (
+                SELECT t.tasks_metadata_hash FROM tasks AS t
+                WHERE t.tasks_task_status = 'pending' AND NOW() > t.tasks_next_try_at
+                ORDER BY t.tasks_last_modified_at ASC
+                FOR UPDATE SKIP LOCKED
+                LIMIT ",
+        );
         query_builder.push_bind(tasks_count);
 
-        // skip locked not to intersect with synchronizer work
         query_builder.push(
-            " FOR UPDATE SKIP LOCKED
-            )
+            ")
             UPDATE tasks t
-            SET tsk_status = 'running',
-            tsk_locked_until = NOW() + INTERVAL '90 seconds'
-            FROM cte
-            WHERE t.tsk_id = cte.tsk_id
-            RETURNING t.tsk_metadata_url, t.tsk_status, t.tsk_attempts, t.tsk_max_attempts;",
+            SET
+                tasks_next_try_at = NOW() + INTERVAL '1 day'
```

Comment:
Am I getting it right, that on selecting a file to download we'll immediately mark it as untouchable for 24 hours? That seems to be a bit too tough. I'd suggest either skipping this update altogether or setting the next_try in some short time. Either option should be ok if we read tasks in a single thread. If in multiple, the later case is more appropriate. Wdyt?